### PR TITLE
release: 4.0

### LINKED ISSUE #292: node 10 support & the future
node 10 will become increasingly hard to support as time goes on. [`ava`](https://github.com/avajs/ava/releases/tag/v4.0.0-alpha.1) and [`eslint`](https://eslint.org/blog/2021/06/whats-coming-in-eslint-8.0.0) have both signaled they will drop support in their next major versions, and `ts-node`'s current major version has (apparently) irreconcilable incompatibilities in its ESM and CJS support.

all of this is motivated by a broad desire by the ecosystem to reach an ESM-only world, and indeed it seems we're on the cusp of entering it. while a laudable goal, it doesn't really make a difference to this library as none of its features are locked into a module format; in fact, this means that to enter the ESM-only world, we would need to strictly weaken `emailjs`'s reach _regardless of future developments in node itself_.

unfortunately, if we wish to keep up with ecosystem enhancements, it means cutting that rope. the choice essentially becomes: do we lock down our dev dependencies at the moment they become incompatible with node 10, effectively locking the library into a slow freeze; or do we keep the future open, at the cost of our entire CJS story?

### PR DESCRIPTION
**Pull request by zackschuster**
- `node` engines entry changed to `>=12`
- minimum supported `typescript` changed to `4.3.5`
- cjs bundle removed; only esm bundle now exported
- tests always run through bundle by default
    - can force it to run through ts files by deleting `email.js` and the `pretest` script 
- `SMTPClient#send` and `SMTPClient#sendAsync` types re-strengthened

closes #292

### CODE REVIEW

File: email.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
-export * from './smtp/address';
-export * from './smtp/client';
-export * from './smtp/connection';
-export * from './smtp/date';
-export * from './smtp/error';
-export * from './smtp/message';
-export * from './smtp/mime';
-export * from './smtp/response';
+export * from './smtp/address.js';
```
Comment (eleith) [inline email.ts:1]: can't this be avoided, and files are referred to without the extension? i find it confusing to refer to a file that doesn't exist (yet)

**Review by eleith — CHANGES_REQUESTED** [review]
changes look good and project is simplified as a result. i have one nit that i would like to find a workaround for though.

File: email.ts (hunk 1 above)
Comment (zackschuster) [reply email.ts:1]: yes, if we specify the `--experimental-module-resolution=node` flag, which iirc is going to be removed when work on loaders is complete (which looks to be soon-ish; [loader chaining is in the PR stage](https://github.com/nodejs/node/pull/42623), which was The Big One as far as features went). 

at any rate, the standard is to specify the extension & typescript's official recommendation (for years) has been ["write the module name as you expect it to work at runtime"](https://github.com/Microsoft/TypeScript/issues/18971#issuecomment-334576547) and just trust the compiler to sort it out.

Comment (eleith) [reply email.ts:1]: on node `v16.4.2` i'm able to run `tsc` and `rollup` without the extensions.

what breaks when the extension is removed?

Comment (zackschuster) [reply email.ts:1]: running the code itself (e.g. testing through the source files & not the bundle)

**Review by eleith — APPROVED** [review]

**Pull request merged**